### PR TITLE
[12.0][FIX] contract: Select the correct sections/notes to be invoiced

### PR DESCRIPTION
**Pull request by ernestotejeda**
Select the correct sections/notes to be invoiced.
For example: Empty contract line sections shouldn't be invoiced.

Cc @Tecnativa TT23706

### CODE REVIEW

File: contract/models/contract.py [hunk 1]
```diff
@@ -412,11 +413,28 @@ def _get_lines_to_invoice(self, date_ref):
...
-            and l.recurring_next_date <= date_ref
-        )
+
+        def can_be_invoiced(l):
+            return (not l.is_canceled and l.recurring_next_date
+                    and l.recurring_next_date <= date_ref)
+
+        res = current_section = self.env['contract.line']
+        add_current_section = add_next_note = False
+        for line in self.contract_line_ids.sorted('sequence'):
```
Comment (pedrobaeza) [inline contract/models/contract.py]: Why do you need `sorted`?

Comment (ernestotejeda) [reply contract/models/contract.py]: If I have contract lines sorted by 'sequence', iterating those lines with a for loop I can assure that lines that go after a certain section correspond to that section and lines inmediatly before a note correspond to that note. With that, i can determine with a single loop, which section/Notes and which product lines can be invoiced.

File: contract/models/abstract_contract_line.py [hunk 2]
```diff
@@ -125,9 +125,27 @@ class ContractAbstractContractLine(models.AbstractModel):
...
-        ('line_note', "Note")], default=False, help="Technical field for UX purpose.")
+    display_type = fields.Selection(
+        selection=[
+            ('line_section', "Section"),
+            ('line_note', "Note"),
+        ],
+        default=False,
+        help="Technical field for UX purpose."
+    )
+    notes_invoicing_mode = fields.Selection(
```
Comment (pedrobaeza) [inline contract/models/abstract_contract_line.py]: Avoid plural

File: contract/models/abstract_contract_line.py [hunk 3]
```diff
@@ -125,9 +125,27 @@ class ContractAbstractContractLine(models.AbstractModel):
...
+            ('with_next_line', 'With next line'),
+            ('custom', 'Custom'),
+        ],
+        default='with_previous_line',
+        help="Defines when the Note is invoiced:\n"
+             "- With previous line: If the previous line can be invoiced.\n"
+             "- With next line: If the previous line can be invoiced.\n"
+             "- Custom: Depending on the recurrence to be define."
+    )
+    is_recurring_note = fields.Boolean(compute="_compute_is_recurring_note")
```
Comment (pedrobaeza) [inline contract/models/abstract_contract_line.py:148]: No need of this extra field as the visibility computation can be done with domain and it's very simple

File: contract/views/abstract_contract_line.xml [hunk 4]
```diff
@@ -8,6 +8,7 @@
         <field name="arch" type="xml">
             <form>
                 <field name="display_type" invisible="1"/>
+                <field name="notes_invoicing_mode" invisible="1"/>
```
Comment (pedrobaeza) [inline contract/views/abstract_contract_line.xml]: No need to add it twice

File: contract/views/abstract_contract_line.xml [hunk 5]
```diff
@@ -62,7 +66,7 @@
                             </div>
                         </group>
                     </group>
-                    <group name="recurrence_info" attrs="{'invisible': [('display_type', '!=', False)]}">
+                    <group name="recurrence_info" attrs="{'invisible': ['|', ('display_type', '=', 'line_section'), ('display_type', '=', 'line_note'), ('note_invoicing_mode', '!=', 'custom')]}">
```
Comment (pedrobaeza) [inline contract/views/abstract_contract_line.xml]:
```suggestion
                    <group name="recurrence_info" attrs="{'invisible': ['|', ('display_type', '=', 'line_section'), '&', ('display_type', '=', 'line_note'), ('note_invoicing_mode', '!=', 'custom')]}">
```

**Comment by pedrobaeza** [comment]
@sbejaoui here we are fixing the problem on the invoiceability of sections and notes, with a good approach IMO:

- Sections are only invoiced if one of the included lines in it is invoiceable.
- Notes can be configured for being invoiceable with the previous or next line, or to be able to define its recurrency independently.

**Review by pedrobaeza — APPROVED** [review]

**Review by carlosdauden — APPROVED** [review]

**Review by sbejaoui — APPROVED** [review]
LGTM, Thanks

File: contract/models/abstract_contract_line.py [hunk 6]
```diff
@@ -125,9 +125,27 @@ class ContractAbstractContractLine(models.AbstractModel):
...
+    note_invoicing_mode = fields.Selection(
+        selection=[
+            ('with_previous_line', 'With previous line'),
+            ('with_next_line', 'With next line'),
+            ('custom', 'Custom'),
+        ],
+        default='with_previous_line',
+        help="Defines when the Note is invoiced:\n"
+             "- With previous line: If the previous line can be invoiced.\n"
+             "- With next line: If the previous line can be invoiced.\n"
```
Comment (sbejaoui) [inline contract/models/abstract_contract_line.py]:
```suggestion
             "- With next line: If the next line can be invoiced.\n"
```

**Comment by sbejaoui** [comment]
@gva-acsone , This should fix the issue we faced as we can now define when the note is invoiced:
- With previous line: If the previous line can be invoiced.
- With next line: If the next line can be invoiced.
- Custom: Depending on the recurrence to be define.

**Comment by pedrobaeza** [comment]
Let's cancel build for changing that text.

**Comment by pedrobaeza** [comment]
Changed now, so let's merge again, but better to put it as minor change, due to fields changing:

/ocabot merge minor

**Pull request merged**